🌐 i18n @ frontend

p-fruck opened the pull request:
<!-- ⚠️ A pull request always needs a related issue! -->
> **As an** administrator
> **I want** a multilingual frontend
> **that** I can understand in my native language

## Description
Added i18n support to frontend (german and english translations)

**Screenshots**
![LoginPageEnglish](https://user-images.githubusercontent.com/30511472/127737705-2f2889db-d71b-4eb5-b98d-e4717dfeecb2.png)
![LoginPageGerman](https://user-images.githubusercontent.com/30511472/127737706-14be7782-eb16-4962-a0fe-9a52f60321fe.png)
![DashboardEnglish](https://user-images.githubusercontent.com/30511472/127737754-665b1157-7604-4f81-ba75-d86c585bb2ff.png)
![DashboardGerman](https://user-images.githubusercontent.com/30511472/127737787-1693b20a-f71f-49e6-be37-227c17ce0807.png)


### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Checklist
<!--- Go over all the following points, and put an `x` in all boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] Make sure you are requesting to main. Don't request to release branch!
- [x] I have linked the related issue.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tjarbo commented:
Great! Many changes :exploding_head: It will take some time from my side

tjarbo commented on packages/frontend/src/components/discord/DiscordPanel.vue:
```diff
@@ -40,21 +41,21 @@ export default {
       this.$store.dispatch('setDiscordChannel', update)
         .then(() => {
-          notifySuccess('Channel auf Discord wurde erfolgreich geändert!');
+          notifySuccess(i18n.t('components.discordPanel.notifications.changedDiscordChannel'));
```
Really like the idea of subsections for notifications :ok_hand:

tjarbo commented on packages/frontend/src/views/Dashboard.vue:
```diff
@@ -1,10 +1,12 @@
   <div id="dashboardview">
-    <TheNavBar></TheNavBar>
+    <TheNavBar>
+      <SwitchLanguage></SwitchLanguage>
```
You have already added `SwitchLanguage` in `TheNavBar.vue`. It shouldn't be required here anymore :thinking: ?

p-fruck replied:
You're right, this was dead code from when I was messing around with nesting components :smile:

tjarbo commented on packages/frontend/tests/unit/views/registration.view.spec.js:
```diff
@@ -6,6 +6,7 @@ import Vuelidate from 'vuelidate';
 import Registration from '@/views/Registration.vue';
 import AuthenticationLayout from '@/layouts/AuthenticationLayout.vue';
 import { v4 } from 'uuid';
+import i18n from '../../../src/i18n';
```
Use `@` instead of `../../../src`
Same comment for all other test files

tjarbo commented on packages/frontend/src/views/Login.vue:
```diff
@@ -1,11 +1,12 @@
 <template>
   <div id="loginview">
+    <SwitchLanguage></SwitchLanguage>
```
Have you tried to add `<SwitchLanguage>` to `layout/AuthenticationLayout` instead to avoid redundant code at the Login and Registration view ?

p-fruck replied:
Didn't think about that, but I like the idea! Done.

tjarbo commented on packages/frontend/src/components/administration/RegistrationTokenModal.vue:
```diff
@@ -2,19 +2,15 @@
-            benutzen, um ein weiteres Gerät zu registrieren (eigener neuer
-            Benutzername ist notwendig!) oder ihn an deine Freunde:innen
-            schicken, damit sie ebenfalls den Service konfigurieren können.
+            {{ $t('components.registrationTokenModal.registrationTokenContent', this.lifetime) }}
```
This translation does not work correctly in both `en` and `de:
![grafik](https://user-images.githubusercontent.com/16938041/127770831-f0b80f93-09ca-4f31-869c-8edd3ea6ef50.png)

p-fruck replied:
Haha, what a shame, I thought it was too trivial to test and forgot the array brackets :laughing:

tjarbo reviewed (changes requested):
Works pretty well! Thank you very much for your contribution :tada: :+1:
I have added some questions and annotations - please have a look :smiley:

p-fruck commented on packages/frontend/src/components/administration/AdministratorList.vue:
```diff
@@ -52,6 +53,7 @@
 import { mapGetters } from 'vuex';
 import { notifySuccess, notifyFailure } from '../../notification';
 import RegistrationTokenModal from './RegistrationTokenModal.vue';
+import i18n from '../../i18n';
```
You're right, I've fixed this in the meantime, but there was one particular issue, that I wasn't able to resolve. In `@/components/statusboard/Status.vue`, I wasn't able to use the `this.$t` syntax within the data block:
```ts
  data: () => ({
    // Column titles
    columns: [
      {
        field: 'key',
        label: i18n.t('general.name'),
      },
      {
        field: 'value',
        label: i18n.t('general.value'),
      },
    ],
```
Since I actually don't know Vue.js very well, do you have any idea how this can be fixed?

tjarbo reviewed (approved):
Perfect :ok_hand: My mistake: Because we are using the language switch twice, the best practices require to remote the "The" from "TheLangaugeSwitch", but yeah ... nvm :sweat_smile: It's fine! 

@p-fruck 
Can you comment a short list of instructions, what steps are required to:
* Add a new language
* Add a missing translation
* Create an overview/results of the current translations?
* Fix a missing translation

p-fruck commented:
> My mistake: Because we are using the language switch twice, the best practices require to remote the "The" from "TheLangaugeSwitch", but yeah ... nvm :sweat_smile:

As we like to complain in Germany: "Einmal mit Profis schaffen :unamused:"    :stuck_out_tongue: 

Should we add something like an i18n quickstart guide to our developer wiki?

# i18n quickstart - frontend

The below content describes the usage of the vue-i18n plugin in our bots frontend. All of the files and commands mentioned below are relative to the `packages/frontend/` folder.

All of our translations are located inside `src/locales/` and use the JSON document structure.

The JSON files are following the same structure as our code directories, which is why the top level entries are `components`,  `layouts` and `views`, which contain the names of the files within the directory, eg `administratorList`. Those entries, which represent our Vue files, contain another set of keys, which is labelled after their function within their parent components. To make this a bit more clear: We have a _component_ _AdministratorList.vue_ which has a _button_, that creates a new registration. The translation key of this button results in `components.administratiorList.createNewRegistrationTokenButton`. So as you can see, we try to use a descriptive approach on our translations, so we can quickly find out their usecase while making them easily translatable.

There is also a `general` section, which contains more generic words like `update`, which are used by multiple components. Notifcations get a separate key within their parent, eg the `Access Expired` message has the translation key `general.notifications.accessExpired`.

## Using the translate function

If you want to add a translate function to your Vue code, there are two methods you have to differentiate:

* Inside a template
  Inside a vue html template, you can simply use the following syntax to translate some key:
  ```html
  <p>{{ $t('general.someTranlsation') }}</p>
  ```
  When binding to a label, you have to use vue-binds:
  ```html
  <input :label="$t('component.example.someLabel')"></input>
  ```
* Inside js code
  Inside you js code, you can use the translation by invoking
  ```js
  this.$t('general.someTranslation');
  ```

  For some scenarios like vue tables, this doesn't work, so you have to import and use i18n manually:
  ```js
  import i18n from '@/i18n';
  ...
  i18n.t('general.someTranslation');
  ```

## Adding a new language

The easiest way to add a new language, is to copy our english translation file (`src/locales/en.json`) to a new JSON file in the same directory with your languages shortcode, eg. `de.json` for german, etc.

After you added the new translation file, you need to translate all of it's content. It's preferred to have native speakers of the given language to take a look at the translation, to mitigate incorrect translations that might occur when using some translation program.

The last required step is to modify `src/components/TheLanguageSwitch.vue`, to add your translation to the switch. Inside the data method, there is an array called `langs`, where you have to add your language to, according to the given schema:
```js
data: () => ({
  langs: [
    { key: 'en', name: 'English' },
    { key: 'de', name: 'Deutsch' },
  ],
}),
```

## Retrieve overview of current translation state

Vues i18n library has an integrated overview of your current translation state. To use it, simply run
`npm run i18n:report`.

If your language file is missing a translation, you'll be reported with the following overview:

```
┌────┬────────────┬────────────────────────────────────────┬────────┬──────────────────────────────┐
│ #  │ Language   │ File                                   │ Line   │ Missing i18n Entry           │
├────┼────────────┼────────────────────────────────────────┼────────┼──────────────────────────────┤
│ 0  │ fr         │ …administration/AdministratorList.vue  │ 16     │ components.administratorLis… │
└────┴────────────┴────────────────────────────────────────┴────────┴──────────────────────────────┘
```
which shows you, that the file `src/components/administration/AdministratorList.vue` is missing the translation used in line 16, which is indeed the key that I have removed from the json file in this example.

If you have an unused entry in one of the translation files, this will also be reported by the command. To simulate this, I removed the logout button translation from the navbar, which resultet in the following statement:

```
┌────┬────────────┬────────────────────────────────────────┬──────┬───────────────────────────────────┐
│ #  │ Language   │ File                                   │ Line │ Unused i18n Entry                 │
├────┼────────────┼────────────────────────────────────────┼──────┼───────────────────────────────────┤
│ 0  │ de         │ /src/locales/de.json                   │ 47   │ components.theNavBar.logoutButton │
│ 1  │ en         │ /src/locales/en.json                   │ 47   │ components.theNavBar.logoutButton │
└────┴────────────┴────────────────────────────────────────┴──────┴───────────────────────────────────┘
```

If everything is fine, you'll be presented with two empty tables:

```
┌────┬────────────┬────────────────────────────────────────┬────────┬──────────────────────────────┐
│ #  │ Language   │ File                                   │ Line   │ Missing i18n Entry           │
└────┴────────────┴────────────────────────────────────────┴────────┴──────────────────────────────┘
┌────┬────────────┬────────────────────────────────────────┬──────┬───────────────────┐
│ #  │ Language   │ File                                   │ Line │ Unused i18n Entry │
└────┴────────────┴────────────────────────────────────────┴──────┴───────────────────┘
```

tjarbo commented:
@p-fruck awesome! I will add this to the docs. Is this all by yourself or a text-template from any website?

p-fruck commented:
@tjarbo I've written it down without the use of any template :wink:
